File: scenarios/dart/dart-binary-parser/template/test/parser_test.dart

import 'dart:typed_data';
import 'package:test/test.dart';
import '../../src/parser.dart'; // Depending on layout, might need adjustment

// Helper to create header
Uint8List createHeader(int type) {
  return Uint8List.fromList([0xAF, 0x01, type]);
}

Uint8List withChecksum(List<int> data) {
  int sumVal = 0;
  for (var b in data) {
    sumVal += b;
  }
  sumVal = sumVal % (1 << 32); // 2^32
  
  final bd = ByteData(4);
  bd.setUint32(0, sumVal, Endian.big);
  
  final b = BytesBuilder();
  b.add(data);
  b.add(bd.buffer.asUint8List());
  return b.toBytes();
}

void main() {
  test('Login Packet', () {
    final header = createHeader(1);
    final userLen = 5;
    final username = 'alice'.codeUnits;
    final hash = List.filled(32, 0xAA);
    
    final payload = BytesBuilder();
    payload.add(header);
    payload.addByte(userLen);
    payload.add(username);
    payload.add(hash);
    
    final packet = withChecksum(payload.toBytes());
    final res = parsePacket(packet);
    
    expect(res['type'], equals('Login'));
    expect(res['username'], equals('alice'));
  });
  
  test('Order Packet', () {
    // ID=123, Symbol=AAPL, Side=1, Price=150.5, Qty=100
    final header = createHeader(2);
    final body = ByteData(25); // 8+4+1+8+4
    body.setInt64(0, 123, Endian.big);
    // Symbol AAPL
    body.setUint8(8, 'A'.codeUnitAt(0));
    body.setUint8(9, 'A'.codeUnitAt(0));
    body.setUint8(10, 'P'.codeUnitAt(0));
    body.setUint8(11, 'L'.codeUnitAt(0));
    
    body.setInt8(12, 1);
    body.setFloat64(13, 150.5, Endian.big);
    body.setUint32(21, 100, Endian.big);
    
    final payload = BytesBuilder();
    payload.add(header);
    payload.add(body.buffer.asUint8List());
    
    final packet = withChecksum(payload.toBytes());
    final res = parsePacket(packet);
    
    expect(res['type'], equals('Order'));
    expect(res['symbol'], equals('AAPL'));
    expect(res['price'], equals(150.5));
  });
  
  test('Underflow', () {
    expect(() => parsePacket([0xAF]), throwsException);
  });
}

File: scenarios/ts-binary-parser/template/tests/parser.test.ts

import { describe, expect, it } from 'vitest';
import { parsePacket } from '../src/parser';

describe('Binary Parser', () => {

    function createHeader(type: number): Buffer {
        return Buffer.from([0xAF, 0x01, type]);
    }

    function calculateChecksum(buf: Buffer): number {
        let sum = 0;
        for (const byte of buf) sum += byte;
        return sum >>> 0; // uint32
    }

    function withChecksum(buf: Buffer): Buffer {
        const sum = calculateChecksum(buf);
        const checksumBuf = Buffer.alloc(4);
        checksumBuf.writeUInt32BE(sum, 0);
        return Buffer.concat([buf, checksumBuf]);
    }

    it('parses Login packet', () => {
        const header = createHeader(0x01);
        const username = Buffer.from('alice');
        const uLen = Buffer.from([username.length]);
        const hash = Buffer.alloc(32).fill(0xAA);

        const payload = Buffer.concat([header, uLen, username, hash]);
        const packet = withChecksum(payload);

        const result = parsePacket(packet);

        if (result.type !== 'Login') throw new Error('Wrong type');
        expect(result.username).toBe('alice');
        expect(result.hash).toBe(hash.toString('hex'));
    });

    it('parses Order packet', () => {
        const header = createHeader(0x02);
        const id = Buffer.alloc(8); id.writeBigInt64BE(123456789n);
        const symbol = Buffer.from('AAPL');
        const side = Buffer.from([1]); // Sell
        const price = Buffer.alloc(8); price.writeDoubleBE(150.50);
        const qty = Buffer.alloc(4); qty.writeUInt32BE(100);

        const payload = Buffer.concat([header, id, symbol, side, price, qty]);
        const packet = withChecksum(payload);

        const result = parsePacket(packet);
        if (result.type !== 'Order') throw new Error('Wrong type');
        expect(result.id).toBe(123456789n);
        expect(result.symbol).toBe('AAPL');
        expect(result.side).toBe('Sell');
        expect(result.price).toBe(150.50);
        expect(result.quantity).toBe(100);
    });

    it('throws on invalid magic byte', () => {
        const buf = Buffer.from([0x00, 0x01, 0x03]); // Wrong magic
        expect(() => parsePacket(buf)).toThrow(/magic/i);
    });

    it('throws on invalid checksum', () => {
        const header = createHeader(0x03); // Heartbeat
        const time = Buffer.alloc(8);
        const payload = Buffer.concat([header, time]);
        const packet = withChecksum(payload);

        // Corrupt last byte
        packet[packet.length - 1] ^= 0xFF;

        expect(() => parsePacket(packet)).toThrow(/checksum/i);
    });

    it('throws on underflow', () => {
        const header = createHeader(0x01);
        expect(() => parsePacket(header)).toThrow(/underflow|short/i);
    });
});
